### ISSUE TITLE
Training about LRS2 dataset

### ISSUE THREAD

**Issue by Physicsmile** [KAIST-AILab/SyncVSR#33]
Hello, when I use your code training on the LRS2 dataset, I find that it's hard to converge. However, I use the code and config you released and don't make any modifications. I use the 'num_training_steps' written in your config which is 450,000. So far, I have trained for 300 epochs, but it still hasn't converged and the decoder_acc_val is not good(around 0.7). But I find you said you only train 200 epochs. So I wonder why this is the case.
Sorry to bother you, looking forward to your reply!

<img width="1350" alt="Image" src="https://github.com/user-attachments/assets/e0bffe9f-5068-41f3-8ac6-df709194f597" />
<img width="1375" alt="Image" src="https://github.com/user-attachments/assets/e5bb783f-a7a2-4ca0-abe9-81a5817e2962" />

**Comment by snoop2head** [comment]
Hello @Physicsmile !
Thank you for filing the issue. Before we begin, may I know the number of GPUs and total batch size?

**Comment by Physicsmile** [comment]
> Hello [@Physicsmile](https://github.com/Physicsmile) ! Thank you for filing the issue. Before we begin, may I know the number of GPUs and total batch size?
Thank you for your reply!
I use eight A6000 GPUs, and batch size=16(same as your batch size in lrs2.yaml). Actually, except set 'resume_from_checkpoint' empty, I don't change other settings in lrs2.yaml.

**Comment by snoop2head** [comment]
@Physicsmile  
Mind if I take a look at the preprocessed video examples and specifics of the loss (loss_ctc_val, loss_att_val, or loss_audio_val)?

**Comment by Physicsmile** [comment]
Sorry I don't record loss_audio_val(same as the released code), but other loss show as below:

<img width="1345" alt="Image" src="https://github.com/user-attachments/assets/192ec5ff-7d87-43f3-9cac-5d3ceef5a4f3" />
<img width="1336" alt="Image" src="https://github.com/user-attachments/assets/f5584a19-c9ec-44ff-b191-8786e57539e2" />

It seems overfitting, but it appears overfitting before 200epoch. So I just wait decoder_acc_val to grow more, because it seems not good now(300 epochs).

And the preprocessed video examples information as follows:

<img width="360" alt="Image" src="https://github.com/user-attachments/assets/e414e6d5-90f8-440b-a8e3-69c5ba46d40a" />
<img width="1068" alt="Image" src="https://github.com/user-attachments/assets/a8e7a4be-ab66-4fc1-ad10-c3fc6dc40d30" />

I use /LRS/video/preprocess/prepare_LRS2.py to preprocess LRS2, not change other things.

Sorry to bother you

**Comment by snoop2head** [comment]
@Physicsmile I am actually thankful that you are showing interest in our work 🤗

- Including LRS2, we train sentence-level VSR models for 100 epochs only (as it says in our paper). I think you don't have to run more than that.
- It seems like the data preprocessing is fine.
- `resume_from_checkpoint` on the config `lrs2.yaml` file is for the evaluation purpose, so you are right about making that part empty

But may I know why the training loss is too low...? Because in case of most of our runs, the `loss_step` hardly drops below 30. How does the trend of `loss_audio_step` look like?

**Comment by Physicsmile** [comment]
I'm extremely grateful for your reply.

I can't reach the results in the paper so I just look almost issues and find you said training 200 epochs on LRS2, but forget what said in the paper, sorry for that.
I trained 70 epochs several days ago, and also remained top-10 ckpt to averge and used to test, but even so I just got 39.17% WER on LRS2. So I wonder if is the num of epoch is not enough. I remember you said in one of the issues that you can reach more than 78 on `decoder_acc_val `during traning, but I can't reach that high. Just as the picture above, I get 67% on `decoder_acc_val` if I just train models for 100 epochs and it is highly probable that perform poorly on LRS2-test based on my experience.

As for the training loss, if I just train 100 epochs, it is around 20 to 25( `loss_step` above is the results of training for 300 epochs, so it is lower than 100 epochs).

`loss_audio_step`is shown as follows:

<img width="1357" alt="Image" src="https://github.com/user-attachments/assets/604fa3cf-1b13-4aec-afc1-cc6b2641ac38" />

**Comment by snoop2head** [comment]
@Physicsmile 
I see. Except for the audio reconstruction loss, the current training specifics exactly follow [AutoAVSR's "3.3. Implementation details"](https://arxiv.org/abs/2303.14307) except for otherwise stated in [our paper](https://arxiv.org/pdf/2406.12233). Maybe you can take a look into that?

**Comment by Physicsmile** [comment]
Thanks for the reminder. I'm also currently conducting the training of AutoAVSR at the same time.  And I still have some I  doubts about SyncVSR at present, and I hope to get them answered:

1. You train sentence-level VSR models for 100 epochs, but I am confused that with `num_training_steps` written in your config which is 450,000. When I use this config, 450,000 steps is more than 100 epochs, so I wonder how to set training epochs to 100 accurately.
2. I have just tried to test the above trained model on LRS2 test, but it only get 39.71% WER 😭(I even choose top-10 ckpt based on  `decoder_acc_val` to average them and test, instead of SyncVSR orginal code choosing the best one to test). I don't know why. May I ask your model trained only on LRS2, so that I can take inference to check if there are any problems on my dataset? 
3. May I ask the test WER on LRS2 of using wav2vec2.0 as quantified methods(because I have used wav2vec2.0 for training to avoid fairseq)? I can't find that on paper or issues.
4. Compared with AutoAVSR, I find that your preprocessed method is a little different from it. **On the one hand**, AutoAVSR first cuts the extremely long pre-training samples into segments with less than 600 frames. Then, during the training process, it will only read these already cut segments.  SyncVSR randomly cut pre-train samples into segments during training. **On the other hand**, AutoAVSR take 96x96 size of video samples as inputs and randomly crop to 88x88, while SyncVSR take 128x128 size of samples and randomresizecrop to 96x96. **And** there are also some change about sampling strategy and configure_optimizers. So may I ask the considerations for your making these changes and whether these changes will have a certain impact on the results?

Since I'm really interested in your work, I'd like to discuss more with you. I would be extremely grateful if my questions could be answered.  🌹🌹🌹

**Comment by snoop2head** [comment]
@Physicsmile 

### Training Epochs
Apologies for the earlier confusion. **To clarify, our experiments on LRS2 and LRS3 actually ran for 150 epochs, not 100 as previously stated.** The reference to 100 epochs applies specifically to the joint pretraining on LRS2 + LRS3 + Vox2, which is part of a separate set of experiments we included in the paper.

### Wav2Vec2.0 Results on LRS2
Our ablations involved:
- A different training framework (JAX/Flax),
- A simpler model architecture (vanilla Transformers vs. Conformer),
- A larger global batch size (128),
- Training on different hardware (TPUs),
- And notably, training *without* a CTC loss component.

Although we observed significant performance improvements of our method over the baseline under those conditions, these differences (especially the omission of CTC loss) meant the results were not directly comparable to our main experiments, so we did not include them in the paper. **Unfortunately, we don’t have an official reported WER for a pure Wav2Vec2.0-based setup on LRS2 under the same conditions. We mainly stuck with vq-wav2vec, so I am unsure of how would it be like in your scenario.**

### Similarities with AutoAVSR
You are correct about using a single best checkpoint for evaluation. Unfortunately, I cannot pinpoint the exact reason for the performance discrepancy. I suggest revisiting the training scheme described in AutoAVSR—for instance, ensuring the visual frontend is set up as instructed in that paper—to see if it aligns with your implementation.

### Differences from AutoAVSR Preprocessing
You are correct about the key differences in our approach:
- Segment Cutting Strategy: Rather than pre-splitting segments to fewer than 600 frames, we randomly sample segments during training. **Our rationale is that the learned positional embeddings can take advantage of this random starting frame.**
- Input Cropping/Resizing: We apply a random-resize-crop from 128×128 down to 96×96, whereas AutoAVSR first uses 96×96 frames and then randomly crops to 88×88. In our internal experiments, particularly on LRW, beginning with a slightly larger region of interest (128×128) and then randomly resizing led to performance gains.

**I hope this helps clear up any confusion!** If you have more questions or want to discuss further details—be it training or preprocessing—feel free to let me know. I’m happy to dive deeper into any part of the pipeline.

Thanks again for your interest!

**Comment by Physicsmile** [comment]
@snoop2head 
Thanks for your patient reply. After carefully reading, I clear up some confusions and still have something uncertain: 
1. Since the released config(`lrs2.yaml`) only uses  `num_training_steps` to control training time, which is set to 450,000(more than 400 epochs actually), so I still don't know how to set training epochs to certain number(such as 150 epochs). So May I ask the` lrs2.yaml` using for training(since the released one is used for evaluation as you mentioned above)?
2. After carefully check about my code, I think there should be no problem with my code because I only use the released SyncVSR code. I also revisit AutoAVSR and I find visual frontend is same as the implementation in SyncVSR on LRS2, So I think I have already aligned with both AutoAVSR and SyncVSR. The only change is that I change the loggers from Wandb to TensorBoard, but I think this change won't cause any influence on the model.  So I guess the only problem may be caused by my dataset. I hope get the released SyncVSR ckpt trained only on LRS2 so that I can inference on my dataset to check if my data is wrong. Could you please upload this model weights?

**Comment by Physicsmile** [comment]
@snoop2head 
Hello! I wonder if it would be convenient for you recently. I would greatly appreciate your response at your earliest convenience.🌹🌹🌹

**Comment by snoop2head** [comment]
@Physicsmile 

### Regarding the batch size and training steps
Our implementation relies on PyTorch Lightning’s [DDP (Distributed Data Parallel)](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_intermediate.html#distributed-data-parallel) backend. We used an effective batch size of 64, derived from a per-device batch size of 16 across 4 devices, with a base learning rate of 1e-3.

I recently noticed that the released version omitted an important detail: repeated sampling. LRS2 contains 140K utterances, but we performed repeated sampling for clips longer than 160 frames (the maximum length in the training split). I've now added repeated sampling—thank you for bringing this to my attention. You can see the relevant code here:

https://github.com/KAIST-AILab/SyncVSR/blob/1fe0ffa27ecc70a878530c8fc11e04482edf508c/LRS/video/datamodule/data_module.py#L62-L68

Regarding your mention of 400 epochs, I’m not entirely sure how that figure was obtained. Here’s our calculation: with repeated sampling, we end up with approximately 190K samples. At an effective batch size of 64, one epoch corresponds to around 3,000 steps. Therefore, training for 150 epochs would involve roughly 450K steps in total. 

### AutoAVSR implementation
I think your data is fine if you were to use the same preprocessing script. But please modify inference config `lrs2.yaml` file for proper training. Refer to the [AutoAVSR](https://arxiv.org/pdf/2303.14307)'s training guide on `3.3. Implementation details` and [resnet18_dctcn_video](https://github.com/mpc001/Lipreading_using_Temporal_Convolutional_Networks?tab=readme-ov-file#model-zoo) for frontend module.

I think I gave sufficient information for replication, but if you have more questions or want to discuss further details feel free to let me know. I’m happy to dive deeper into any part of the pipeline.

Thanks again for your interest!

**Comment by Physicsmile** [comment]
@snoop2head 

Thanks for your reminder. I'm training on the newly released version and use four GPUs same as you mentioned, it will take 30 minutes every epoch so I have to wait several days to check the results. Would you mind not closing this issue for now? I hope to have further discussions with you later.
Meanwhile, I still have some doubts:

1. I have checked the maximum length in the LRS2 training split, it is 154, while the max length in `./datamodule/video_length.npy` is 155 which is used to distinguished from pretrain samples. And your newly released code takes 160 as the maximum length to repeated sampling. So I'm confused about these numbers. Is 160 chosen to be the approximate number for rounding？
2. I revisit [AutoAVSR](https://arxiv.org/pdf/2303.14307)'s training guide on `3.3. Implementation details`, it said "For our audio- and visual-only ASR models, we use a ResNet- based front-end module pre-trained on LRW". So I wonder if the frontend of SyncVSR is initialised from the model pre-trained on LRW([resnet18_dctcn_video](https://github.com/mpc001/Lipreading_using_Temporal_Convolutional_Networks?tab=readme-ov-file#model-zoo)) then trained on LRS2.
3. The structure of frontend module I used is same as SyncVSR, and it is also same as [resnet18_dctcn_video](https://github.com/mpc001/Lipreading_using_Temporal_Convolutional_Networks?tab=readme-ov-file#model-zoo)(AutoAVSR), So I think there is no problem with my model.

**Comment by snoop2head** [comment]
@Physicsmile 

**Without our audio reconstruction loss, did you replicate AutoAVSR or CM-Aux baseline?** Does the audio reconstruction loss diminish the baseline performance, or does it enhance it? These are important questions to address. AutoAVSR is an extension of [Visual Speech Recognition for Multiple Languages in the Wild](https://github.com/mpc001/Visual_Speech_Recognition_for_Multiple_Languages) (=CM-Aux) but training more on synthetically generated transcripts. I think the current performance you've reported is below than that of such baseline, so I’m uncertain how best to assist in this context. 

When replicating the baseline without audio reconstruction loss, I recommend referring to the `Experimental settings` of [End-to-end Audio-visual Speech Recognition with Conformers](https://arxiv.org/pdf/2102.06657) (=CM-Seq2Seq) which is a setup for both CM-Aux and AutoAVSR. Check if you can replicate the performance of CM-Aux baseline.

Regarding the recent change, please consider it a hotfix. Previously, I saved the created file paths from self.train_filenames as lrs2.txt or lrs3.txt and reloaded them. It should be straightforward for you to adapt this approach for LRS2 experiments. However, since we need to accommodate various combinations—LRS2, LRS3, LRS2+LRS3, and LRS2+LRS3+VoxCeleb—it will take some time to finalize a comprehensive solution. Please feel free to make any necessary modifications on your end before I release a more complete update. Lastly, as you may have noticed, the code uses `np.ceil()`. The slight discrepancy—like handling 156 frames as 160 frames once or twice—shouldn’t significantly affect your results.

**Comment by Physicsmile** [comment]
@snoop2head 
Actually, I have already conducted several experiments on both AutoAVSR and SyncVSR. 
To validate the influence of audio reconstruction loss, I tried to train them from scratch with same input data and config. But the results is SyncVSR had a poor WER(about 40), while AutoAVSR could not converge.
I noticed AutoAVSR said trained from scratch results in poor performance so it used curriculum learning. Then I used the dataset preprocess of AutoAVSR and tried  curriculum learning(Stage1:4s;Stage2:24s) on both AutoAVSR and SyncVSR. If models were not initialised from the model pre-trained on LRW(only change learning rate of two stages), SyncVSR test WER is around 33% and AutoAVSR is 35%. If model was initialised from the model pre-trained on LRW, AutoAVSR can reach 34%. As the paper reported SyncVSR WER on LRS2 test is 30.7%, so I wonder if SyncVSR has same strategy(use pre-trained frontend) to train on LRS2.

**Comment by snoop2head** [comment]
@Physicsmile 
In the `Training Recipe` section, our paper specifies that "The rest of the training specifics follow the previous work’s settings from AutoAVSR". Add audio reconstruction loss on top of AutoAVSR, which is equivalent to CM-Aux, and check if it improves it or not. Refer to the existing work if your question were to line up with the corresponding paper.

**Comment by Physicsmile** [comment]
@snoop2head 
So SyncVSR need to load the pre-trained frontend when training on LRS2?

**Comment by Physicsmile** [comment]
@snoop2head 
Hello, 
I retrain SyncVSR on hotfix version, with same number of GPUs(4 GPUs), same configs(lrs2.yaml), same dataset(LRS2 preprocess) and same code(hotfix version), but I get 53.54% WER on LRS2 test. Some specifics of the loss show as below:

<img width="1040" alt="Image" src="https://github.com/user-attachments/assets/460d9667-7632-439c-b5bb-a6dc0c8ee13f" />
<img width="1051" alt="Image" src="https://github.com/user-attachments/assets/9d9dda50-6df9-42ac-8cf9-0a987059f40c" />
<img width="1048" alt="Image" src="https://github.com/user-attachments/assets/850c65a9-b844-495f-bcea-7e255486d2eb" />
<img width="1053" alt="Image" src="https://github.com/user-attachments/assets/3c139eef-7cf8-4238-a8d3-df1131652b7b" />
Previously, I used 8 GPUs and got 39% WER while this time get even worse. So I'm not sure if I should load the pre-trained frontend to match the result reported in paper.

**Comment by HarukiK2** [comment]
Hi @snoop2head, @Physicsmile,

Thank you for the informative and detailed discussion above.
I’m currently trying to reproduce the SyncVSR results on the LRS2 dataset using the official code (including the recent hotfix), and I am also facing similar issues with poor convergence and high WER.

**My Setup**:
GPUs: 4xA100

**Batch size**: 16 / 1GPU (effectively 64)

Training steps/config: lrs2.yaml (official release, no modification except logger, codec)
**Dataset**: LRS2 processed with prepare_LRS2.py as instructed
**My Observations**:
After training to 150 epochs (approx. 450K steps), the model’s performance remains poor.

I observed WER around 50% on LRS2 test set, which is significantly worse than the reported 30.7%.

I’m wondering whether a pre-trained visual frontend on LRW is necessary to reach the reported performance on LRS2?
Also, do you recommend curriculum learning (e.g., staged training) as a requirement for convergence in your recent experiments?

Any guidance or clarification would be greatly appreciated, especially on:

・Whether the released weights for LRS2-only training can be shared for comparison
・Expected decoder accuracy trends during training on LRS2 (with/without pretraining)
・Whether audio reconstruction loss helps or hinders in the case of LRS2-only training

Thanks again for your impressive work and for supporting the community with such detailed replies.
Looking forward to your thoughts!

Best regards,
HarukiK2
